sw/airborne/modules/guidance/gvf_parametric/gvf_parametric.cpp

```diff
@@ -0,0 +1,455 @@
+{
+  gvf_parametric_control.s = s;
+}
+
+void gvf_parametric_control_2D(float kx, float ky, float f1, float f2, float f1d, float f2d, float f1dd, float f2dd)
+{
+
+  uint32_t now = get_sys_time_msec();
+  gvf_parametric_control.delta_T = now - gvf_parametric_t0;
+  gvf_parametric_t0 = now;
+
+  if (gvf_parametric_control.delta_T > 300) { // We need at least two iterations for Delta_T
+    gvf_parametric_control.w = 0; // Reset w since we assume the algorithm starts
+    return;
+  }
+
+  float L = gvf_parametric_control.L;
+  float beta = gvf_parametric_control.beta*gvf_parametric_control.s;
+
+  Eigen::Vector3f X;
+  Eigen::Matrix3f J;
+
+  // Error signals phi_x and phi_y
+  struct EnuCoor_f *pos_enu = stateGetPositionEnu_f();
+  float x = pos_enu->x;
+  float y = pos_enu->y;
+
+  float phi1 = L * (x - f1);
+  float phi2 = L * (y - f2);
+
+  gvf_parametric_trajectory.phi_errors[0] = phi1; // Error signals for the telemetry
+  gvf_parametric_trajectory.phi_errors[1] = phi2;
+
+  // Chi
+  X(0) = L*beta*f1d - kx*phi1;
+  X(1) = L*beta*f2d - ky*phi2;
+  X(2) = L + beta*(kx*phi1*f1d + ky*phi2*f2d);
+  X *= L;
+
+  // Jacobian
+  J.setZero();
+  J(0, 0) = -kx*L;
+  J(1, 1) = -ky*L;
+  J(2, 0) = (beta*L)*(beta*f1dd+kx*f1d);
+  J(2, 1) = (beta*L)*(beta*f2dd+ky*f2d);
+  J(2, 2) = beta*beta*(kx*(phi1*f1dd-L*f1d*f1d) + ky*(phi2*f2dd-L*f2d*f2d));
+  J *= L;
+
+  // Guidance algorithm
+  float ground_speed = stateGetHorizontalSpeedNorm_f();
+  float w_dot = (ground_speed * X(2)) / sqrtf(X(0) * X(0) + X(1) * X(1));
+
+  Eigen::Vector3f xi_dot;
+  struct EnuCoor_f *vel_enu = stateGetSpeedEnu_f();
+  float course = stateGetHorizontalSpeedDir_f();
+
+  xi_dot << vel_enu->x, vel_enu->y, w_dot;
+
+  Eigen::Matrix3f G;
+  Eigen::Matrix3f Gp;
+  Eigen::Matrix<float, 2, 3> Fp;
+  Eigen::Vector2f h;
+  Eigen::Matrix<float, 1, 2> ht;
+
+  G << 1, 0, 0,
+       0, 1, 0,
+       0, 0, 0;
+  Fp << 0, -1, 0,
+        1,  0, 0;
+  Gp << 0, -1, 0,
+        1,  0, 0,
+        0,  0, 0;
+
+  h << sinf(course), cosf(course);
+  ht = h.transpose();
+
+  Eigen::Matrix<float, 1, 3> Xt = X.transpose();
+  Eigen::Vector3f Xh = X / X.norm();
+  Eigen::Matrix<float, 1, 3> Xht = Xh.transpose();
+  Eigen::Matrix3f I;
+  I.setIdentity();
+
+  float aux = ht * Fp * X;
+  float heading_rate = -1 / (Xt * G * X) * Xt * Gp * (I - Xh * Xht) * J * xi_dot - (gvf_parametric_control.k_psi * aux / sqrtf(Xt * G * X));
+
+  // Low-level control
+  if (autopilot_get_mode() == AP_MODE_AUTO2) {
+
+    // Virtual coordinate
+    gvf_parametric_control.w += w_dot * gvf_parametric_control.delta_T * 1e-3;
+
+    // Lateral XY coordinates
+    lateral_mode = LATERAL_MODE_ROLL;
+
+    struct FloatEulers *att = stateGetNedToBodyEulers_f();
+
+    h_ctl_roll_setpoint =
+      -gvf_parametric_control.k_roll * atanf(heading_rate * ground_speed / GVF_PARAMETRIC_GRAVITY / cosf(att->theta));
+    BoundAbs(h_ctl_roll_setpoint, h_ctl_roll_max_setpoint); // Setting point for roll angle
+  }
```

Comment:
Indeed, and if one uses the 3D version, the vertical_speed_setpoint (I think that is the name) is also set in the same way.

What would you propose? The best indeed would be to have a #IF firmware blabla, and inside a set function set_v_speed(x), or set_turning_rate(x)?

Reply:
I was thinking to add parameters to function with output values, so that everything in this function is only related to your algo. So something like
```
void gvf_parametric_control_2D(...., float *roll_setpoint)
{
  ...
  * roll_setpoint = ... ;
}
```
They can also be part of the `gvf_parametric_control` structure to avoid extra parameters.

Even the test on AP mode and the `lateral_mode` to ROLL should be left outside.

Then an extra function, called at an upper level, can be used to "apply" the commands to the system with the proper `#if`. Even better, this function is a callback  ? I'm not sure at this point.
Actually, this is a tricky case of guidance function called by flight plan, where you don't want to care about lower level, but don't want (at least me) to be tied to it either.

Reply:
For your two questions: yes.
However, your case has some difficulties, because it should be called from flight plan. So what you want is a single function with "high level" inputs and returning `true` until completion. But to split the guidance algorithm from setting the right stabilization inputs, the best is to have two functions (one generic, one firmware specific).
One reason of this mess is that, as you said, the global variables for low level input are directly set instead of using some abstract interface.
Also, in this particular case, AP generation is not a big help, since you are always in "NAV" mode in the end.
So, options I can see are:
- using `#ifdef` selection at the end of the algorithm part to set the global function depending on the firmware (I don't like this one)
- the algo itself is a generic library and the specific functions use the `#ifdef` option (or call a final factorized function to do that for them)
- we use the possibility to have firmware specific part in the module XML file to have implementation specific to each firmware (see the "mission" modules for examples)
- the pointers to the output variables are passed as parameters when calling the function from the flight plan (not really user friendly, but the most flexible probably)

If it were only me, I would probably go with the last one because being to specify explicitly the inputs and outputs of each control loop, whoever calls it, is were I want to go. But that's clearly not the most convenient for new users...

Reply:
Indeed, from the Flightplan, we call "high level" functions with "descriptive" inputs, such as altitude/radius and things like that.
The algorithm is hidden from the user. 

No problem, I can follow the fourth point, but I do not know exactly how. Do you have any example somewhere? 

Right now, the one that sounds better to me is the second one. The algorithm is generic, and in the end it calls an "execution function". Within that function (even I can put it in a different file), I do only staff related to firmware/hardware so that the algorithm is totally independent of the chosen firmware. Later, once is required because we update pprz or whatever, it should be easier to adapt the code.

Let me know what you think, and I will work on it in September :P

Reply:
You can give a go with the second point. If the firmware specific code, becomes to big, then the third one is almost the same but cleaner.
Fourth option is maybe too tricky for users.